solidity/contracts/hooks/CCIPHookPublisher.sol

```diff
@@ -0,0 +1,217 @@
+
+    // ============ Events ============
+
+    event MessageSent(
+        bytes32 indexed messageId, // The unique ID of the CCIP message.
+        uint64 indexed destinationChainSelector, // The chain selector of the destination chain.
+        address receiver, // The address of the receiver on the destination chain.
+        bytes callData, // The payload being sent
+        address feeToken, // the token address used to pay CCIP fees.
+        uint256 fees // The fees paid for sending the CCIP message.
+    );
+
+    // ============ Storage ============
+
+    mapping(uint64 => bool) public allowlistedDestinationChains;
+    address internal immutable i_router;
+
+    // ============ Constructor ============
+
+    /// @notice Constructor initializes the contract with the router address.
+    /// @param _router The address of the router contract.
+    constructor(
+        address _router,
+        address _mailbox,
+        uint32 _destinationDomain,
+        bytes32 _ism
+    )
+        AbstractMessageIdAuthHook(_mailbox, _destinationDomain, _ism) {
+        i_router = _router;
+    }
+
+    // ============ Modifiers ============
+
+    /// @dev Modifier that checks if the chain with the given destinationChainSelector is allowlisted.
+    /// @param _destinationChainSelector The selector of the destination chain.
+    modifier onlyAllowlistedDestinationChain(uint64 _destinationChainSelector) {
+        if (!allowlistedDestinationChains[_destinationChainSelector])
+            revert DestinationChainNotAllowlisted(_destinationChainSelector);
+        _;
+    }
+
+    // ============ Internal functions ============
+
+    /// @notice Construct a CCIP message.
+    /// @dev This function will create an EVM2AnyMessage struct with all the necessary information for sending a text.
+    /// @param _receiver The address of the receiver.
+    /// @param _callData The string data to be sent.
+    /// @param _feeTokenAddress The address of the token used for fees. Set address(0) for native gas.
+    /// @return Client.EVM2AnyMessage Returns an EVM2AnyMessage struct which contains information for sending a CCIP message.
+    function _buildCCIPMessage(
+        address _receiver,
+        bytes memory _callData,
+        address _feeTokenAddress
+    ) internal pure returns (Client.EVM2AnyMessage memory) {
+        // Create an EVM2AnyMessage struct in memory with necessary information for sending a cross-chain message
+        return
+            Client.EVM2AnyMessage({
+                receiver: abi.encode(_receiver), // ABI-encoded receiver address
+                data: abi.encode(_callData), // ABI-encoded payload
+                tokenAmounts: new Client.EVMTokenAmount[](0), // Empty array aas no tokens are transferred
+                extraArgs: Client._argsToBytes(
+                    // Additional arguments, setting gas limit and non-strict sequencing mode
+                    Client.EVMExtraArgsV1({gasLimit: 200_000, strict: false})
```

Comment:
True, updated

Just in case: according to CCIP best practices, it makes sense to make this variable mutable, so it can be updated in the future, if necessary.